libraries/AC_AutoTune/AC_AutoTune.cpp

```diff
@@ -109,6 +109,51 @@ void AC_AutoTune::stop()
     // we expect the caller will change the flight mode back to the flight mode indicated by the flight mode switch
 }
 
+// Autotune aux function trigger
+void AC_AutoTune::do_aux_function(const RC_Channel::AuxSwitchPos ch_flag)
+{
+    if (mode != TuneMode::SUCCESS) {
+        if (ch_flag == RC_Channel::AuxSwitchPos::HIGH) {
+            gcs().send_text(MAV_SEVERITY_NOTICE,"AutoTune: must be complete to test gains");
+        }
+        return;
+    }
+
+    switch(ch_flag) {
+        case RC_Channel::AuxSwitchPos::LOW:
+            // load original gains
+            load_gains(GainType::GAIN_ORIGINAL);
+            update_gcs(AUTOTUNE_MESSAGE_TESTING_END);
+            break;
+        case RC_Channel::AuxSwitchPos::MIDDLE:
+            // Middle position is unused for now
+            break;
+        case RC_Channel::AuxSwitchPos::HIGH:
+            // Load tuned gains
+            load_gains(GainType::GAIN_TUNED);
+            update_gcs(AUTOTUNE_MESSAGE_TESTING);
+            break;
+    }
+
+    have_pilot_testing_command = true;
+}
+
+// Possibly save gains, called on disarm
+void AC_AutoTune::disarmed(const bool in_autotune_mode)
+{
+    // True if pilot is testing tuned gains
+    const bool testing_tuned = have_pilot_testing_command && (loaded_gains == GainType::GAIN_TUNED);
+
+    // True if in autotune mode and no pilot testing commands have been received
+    const bool tune_complete_no_testing = !have_pilot_testing_command && in_autotune_mode;
+
+    if (tune_complete_no_testing || testing_tuned) {
+        save_tuning_gains();
```

Comment:
do you intend to save_tuning_gains() if switch is in lower position so we currently have the old gains loaded?

Reply:
If you have used the switch to test the gains and you have the switch in the "original gains" position then it should not save the tuned gains.

If you have not used the switch to test the gains and are in AutoTune mode then it will save the gains. The is should be the same as the current behaviour.

Reply:
Sounds confusing.

Reply:
I'm open to better suggestions but I could not think of anything else while preserving the current behaviour, we can't make it a requirement that everyone must setup the new switch to run autotune.

Now we can send aux functions from the gcs we can't really check if the user has setup a switch until they first trigger it.

Reply:
How about reverse the switch low = save, high = test?

Reply:
I'm not sure how that is less confusing, the pattern (there are one or two exceptions) for functions is that high is to take some action. In this case the action is the thing we have added namely the ability to test gains in any mode. Original gains and discarding in other modes is the existing behavior, it would be abit odd to require the user to flip a switch to high for that.

Don't forget that aux functions are edge triggered, it does not matter where the switch is until autotune has completed, only changes after that (and before you disarm) will do anything.